### PR TITLE
server, config: Self protection config

### PR DESCRIPTION
**Pull request by CabinfeverB**
### What problem does this PR solve?
This is a child PR of #4466
For details, please refer to #4466

**NOTE: This PR will not merge.**

### What is changed and how it works?
add config for self protection 

### Check List

Tests

- No code

### Release note

```release-note
None
```

### CODE REVIEW

File: pkg/apiutil/apiutil.go [hunk 1]
```diff
@@ -127,3 +132,56 @@ func ErrorResp(rd *render.Render, w http.ResponseWriter, err error) {
...
+func GetIPAddrFromGRPCContext(ctx context.Context) string {
+	ip, ok := GetRealIPAddrFromGRPCContext(ctx)
+	if ok {
+		return ip
+	}
+	ip = GetPeerAddrFromGRPCContext(ctx)
+	return ip
+}
+
+func GetRealIPAddrFromGRPCContext(ctx context.Context) (string, bool) {
```
Comment (disksing) [inline pkg/apiutil/apiutil.go:146]: Need comment for exported functions.

File: pkg/apiutil/apiutil.go [hunk 2]
```diff
@@ -127,3 +132,56 @@ func ErrorResp(rd *render.Render, w http.ResponseWriter, err error) {
...
+}
+
+// GetIPAddrFromHTTPRequest returns http client IP from context
+func GetIPAddrFromHTTPRequest(r *http.Request) string {
+	ips := strings.Split(r.Header.Get("X-Forwarded-For"), ",")
+	if ips[0] != "" {
+		return ips[0]
+	}
+
+	ip := r.Header.Get("X-Real-Ip")
```
Comment (disksing) [inline pkg/apiutil/apiutil.go:177]: Can we make it consistant? I see `x-real-ip` in `GetRealIPAddrFromGRPCContext`.

File: conf/config.toml [hunk 3]
```diff
@@ -200,3 +200,12 @@
 
 ## When enabled, usage data will be sent to PingCAP for improving user experience.
 # enable-telemetry = true
+
+[self-protection]
```
Comment (disksing) [inline conf/config.toml:204]: Maybe `rate-limit`?

File: pkg/apiutil/apiutil.go (hunk 2 above)
Comment (CabinfeverB) [reply pkg/apiutil/apiutil.go:177]: ok, http also supports lowercase

File: pkg/apiutil/apiutil.go (hunk 1 above)
Comment (CabinfeverB) [reply pkg/apiutil/apiutil.go:146]: comment to be written

File: server/config/config.go [hunk 4]
```diff
@@ -163,6 +163,8 @@ type Config struct {
 	Dashboard DashboardConfig `toml:"dashboard" json:"dashboard"`
 
 	ReplicationMode ReplicationModeConfig `toml:"replication-mode" json:"replication-mode"`
+
+	SelfProtectionConfig SelfProtectionConfig `toml:"SelfProtectionConfig"`
```
Comment (bufferflies) [inline server/config/config.go:167]:
```suggestion
	SelfProtectionConfig SelfProtectionConfig `toml:"self-protection-config"`
```

File: server/config/config.go [hunk 5]
```diff
@@ -1404,3 +1411,19 @@ type SecurityConfig struct {
 	RedactInfoLog bool              `toml:"redact-info-log" json:"redact-info-log"`
 	Encryption    encryption.Config `toml:"encryption" json:"encryption"`
 }
+
+type ComponenetRateLimits struct {
+	Components string `toml:"components"`
```
Comment (bufferflies) [inline server/config/config.go:1416]: why not add consider serial type by json ?

File: server/middleware/self_protection.go [hunk 6]
```diff
@@ -0,0 +1,432 @@
...
+	return func(srv interface{}, stream grpc.ServerStream, info *grpc.StreamServerInfo, handler grpc.StreamHandler) error {
+		if h.SelfProtectionHandleGRPC(info.FullMethod, stream.Context()) {
+			return status.Errorf(codes.ResourceExhausted, "%s is rejected by grpc_ratelimit middleware, please retry later.", info.FullMethod)
+		}
+		return handler(srv, stream)
+	}
+}
+
+// UserSignatureGRPCClientInterceptorBuilder add component user signature in gRPC
+type UserSignatureGRPCClientInterceptorBuilder struct {
```
Comment (bufferflies) [inline server/middleware/self_protection.go:390]: The function signature may be too long.

**Comment by CabinfeverB** [comment]
close this PR to change spliting small PRs

**Pull request closed without merging**